### PR TITLE
d/aws_s3_bucket_object: Don't crash when S3 HeadObject returns empty response

### LINKED ISSUE #14151: Bad key value in s3 data source causes crash
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.12.21
+ provider.aws v2.69.0
+ provider.cloudflare v2.8.0
+ provider.template v2.1.2

### Affected Resource(s)

* aws_s3_bucket_object

### Terraform Configuration Files

```hcl
data "aws_s3_bucket_object" "placeholder" {
  bucket = "some_bucket_name"
  key    = "/" 
}
```

### Debug Output

https://gist.github.com/johnkeates/6f16cce6357fa1c20c16c9077939b996

### Expected Behavior

An error should have been shown, either passed down from the AWS SDK or from the provider as this input can never result in any correct result.

### Actual Behavior

The provider crashed

### Steps to Reproduce

1. Create this:
```
data "aws_s3_bucket_object" "placeholder" {
  bucket = "some_bucket_name"
  key    = "/" 
}
```

2. `terraform apply`

### Factoid
I accidentally came across this because I was using the output of a resource to pipe via interpolation in to this data item, and the resource returned an empty string. Both the empty string and a slash cause this behaviour. While this is not valid use of the data item it should also not crash the provider and cause terraform to get a dropped RPC pipe. A true empty string returns an:

```
Error: Failed getting S3 object: InvalidParameter: 1 validation error(s) found.
- minimum field size of 1, HeadObjectInput.Key.
```

but if your string is "/" it crashes.

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #14151.
Relates #12992.
Relates #7584.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
data-source/aws_s3_bucket_object: Don't crash when S3 `HeadObject` API returns empty response.
```

Output from acceptance testing:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAWSS3BucketObject_' 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccDataSourceAWSS3BucketObject_ -timeout 120m
=== RUN   TestAccDataSourceAWSS3BucketObject_basic
=== PAUSE TestAccDataSourceAWSS3BucketObject_basic
=== RUN   TestAccDataSourceAWSS3BucketObject_basicViaAccessPoint
=== PAUSE TestAccDataSourceAWSS3BucketObject_basicViaAccessPoint
=== RUN   TestAccDataSourceAWSS3BucketObject_readableBody
=== PAUSE TestAccDataSourceAWSS3BucketObject_readableBody
=== RUN   TestAccDataSourceAWSS3BucketObject_kmsEncrypted
=== PAUSE TestAccDataSourceAWSS3BucketObject_kmsEncrypted
=== RUN   TestAccDataSourceAWSS3BucketObject_allParams
=== PAUSE TestAccDataSourceAWSS3BucketObject_allParams
=== RUN   TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOff
=== PAUSE TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOff
=== RUN   TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOn
=== PAUSE TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOn
=== RUN   TestAccDataSourceAWSS3BucketObject_LeadingSlash
=== PAUSE TestAccDataSourceAWSS3BucketObject_LeadingSlash
=== RUN   TestAccDataSourceAWSS3BucketObject_MultipleSlashes
=== PAUSE TestAccDataSourceAWSS3BucketObject_MultipleSlashes
=== RUN   TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
=== PAUSE TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
=== CONT  TestAccDataSourceAWSS3BucketObject_basic
=== CONT  TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOn
=== CONT  TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
=== CONT  TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOff
=== CONT  TestAccDataSourceAWSS3BucketObject_MultipleSlashes
=== CONT  TestAccDataSourceAWSS3BucketObject_allParams
=== CONT  TestAccDataSourceAWSS3BucketObject_LeadingSlash
=== CONT  TestAccDataSourceAWSS3BucketObject_kmsEncrypted
=== CONT  TestAccDataSourceAWSS3BucketObject_readableBody
=== CONT  TestAccDataSourceAWSS3BucketObject_basicViaAccessPoint
--- PASS: TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey (42.12s)
--- PASS: TestAccDataSourceAWSS3BucketObject_basic (45.22s)
--- PASS: TestAccDataSourceAWSS3BucketObject_readableBody (47.39s)
--- PASS: TestAccDataSourceAWSS3BucketObject_allParams (47.41s)
--- PASS: TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOff (48.20s)
--- PASS: TestAccDataSourceAWSS3BucketObject_kmsEncrypted (48.49s)
--- PASS: TestAccDataSourceAWSS3BucketObject_basicViaAccessPoint (48.81s)
--- PASS: TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOn (49.41s)
--- PASS: TestAccDataSourceAWSS3BucketObject_LeadingSlash (67.87s)
--- PASS: TestAccDataSourceAWSS3BucketObject_MultipleSlashes (68.27s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	68.374s
```

### CODE REVIEW

File: aws/internal/keyvaluetags/s3_tags.go [hunk 1]
```diff
@@ -87,6 +87,10 @@ func S3ObjectListTags(conn *s3.S3, bucket, key string) (KeyValueTags, error) {
 
 	output, err := conn.GetObjectTagging(input)
 
+	if awsbase.IsAWSErr(err, "NoSuchTagSet", "") {
```
Comment (ewbankkit) [inline aws/internal/keyvaluetags/s3_tags.go:90]: Prevents errors:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey -timeout 120m
=== RUN   TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
=== PAUSE TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
=== CONT  TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
--- FAIL: TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey (24.58s)
    testing.go:684: Step 0 error: errors during apply:
        
        Error: error listing tags for S3 Bucket (tf-acc-test-2642544354113174819) Object (/): NoSuchTagSet: The TagSet does not exist
        	status code: 404, request id: 3ADF2BE8D6DAD57E, host id: zIz72pHcPivqgl0t79J/9kF02E5N7ibZy3Rof+ggH2+Y3sffOQsjRQVkP5B/PinBmLJnI1nfvCk=
        
          on /tmp/tf-test734233549/main.tf line 6:
          (source code not available)
        
        
    testing.go:745: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: errors during refresh: error listing tags for S3 Bucket (tf-acc-test-2642544354113174819) Object (/): NoSuchTagSet: The TagSet does not exist
        	status code: 404, request id: 9746D2A0387F2F59, host id: tCpkS+zPf8miGLqN8BaDPGbPBs03PZoC3iJHWvhd3SeN7k/Y1BD1AKJmujxgfBqGshau1OP9Pu4=
        
        State: <nil>
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	24.654s
FAIL
GNUmakefile:26: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

Comment (bflad) [reply aws/internal/keyvaluetags/s3_tags.go:90]: After recent merge, ready for rebase. 👍 

```
# github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags
aws/internal/keyvaluetags/s3_tags.go:91:5: undefined: awsbase
```

```suggestion
	if tfawserr.ErrCodeEquals(err, tfs3.NoSuchTagSet) {
```

Comment (ewbankkit) [reply aws/internal/keyvaluetags/s3_tags.go:90]: Will do (plus change the error code constants to include `ErrCode` prefix).

**Comment by ewbankkit** [comment]
Not sure why the PR blew up in size there but I have rebased and now it's back down to `size/S`.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey -timeout 120m
=== RUN   TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
=== PAUSE TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
=== CONT  TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey
--- PASS: TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey (18.81s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	18.888s
```

**Review by bflad — APPROVED** [review]
Looks good to me 🚀 

Output from acceptance testing:

```
--- PASS: TestAccDataSourceAWSS3BucketObject_kmsEncrypted (53.21s)
--- PASS: TestAccDataSourceAWSS3BucketObject_readableBody (53.78s)
--- PASS: TestAccDataSourceAWSS3BucketObject_SingleSlashAsKey (54.16s)
--- PASS: TestAccDataSourceAWSS3BucketObject_basic (55.33s)
--- PASS: TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOff (55.08s)
--- PASS: TestAccDataSourceAWSS3BucketObject_ObjectLockLegalHoldOn (56.63s)
--- PASS: TestAccDataSourceAWSS3BucketObject_allParams (57.27s)
--- PASS: TestAccDataSourceAWSS3BucketObject_basicViaAccessPoint (59.73s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_startAfter (74.51s)
--- PASS: TestAccDataSourceAWSS3BucketObject_MultipleSlashes (76.46s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_fetchOwner (76.46s)
--- PASS: TestAccDataSourceAWSS3BucketObject_LeadingSlash (78.07s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_maxKeys (78.09s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_basicViaAccessPoint (77.91s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_basic (78.31s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_all (79.05s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_prefixes (78.72s)
--- PASS: TestAccDataSourceAWSS3BucketObjects_encoded (78.60s)
```

**Pull request merged**